### PR TITLE
Feature/login conv

### PR DESCRIPTION
**Pull request by meletios**

### CODE REVIEW

File: src/main/java/com/hedvig/botService/chat/OnboardingConversationDevi.kt [hunk 1]
```diff
@@ -363,14 +365,122 @@ constructor(
...
+                        false
+                    ),
+                    SelectOption("Logga in med email och lösenord", MESSAGE_LOGIN_ASK_EMAIL),
+                    SelectOption("Jag är inte medlem", MESSAGE_ONBOARDINGSTART_ASK_EMAIL)
+                )
+            ) { body, uc, message ->
+                body.text = body.selectedItem.text
+                addToChat(message, uc)
+                val obd = uc.onBoardingData
+                if (body.selectedItem.value == "message.bankid.autostart.respond") {
```
Comment (palmfjord) [inline src/main/java/com/hedvig/botService/chat/OnboardingConversationDevi.kt]: maybe we could do a `when` here? something like this

```kt
when (body.selectedItem.value) {
  "message.bankid.autostart.respond" -> { ... }
  MESSAGE_ONBOARDINGSTART_ASK_EMAIL -> { ... }
}
```

File: src/main/java/com/hedvig/botService/chat/OnboardingConversationDevi.kt [hunk 2]
```diff
@@ -1384,6 +1493,34 @@ constructor(
                 endConversation(userContext)
                 return
             }
+            MESSAGE_LOGIN_ASK_EMAIL -> {
+                val trim = m.body.text.trim { it <= ' ' }
```
Comment (palmfjord) [inline src/main/java/com/hedvig/botService/chat/OnboardingConversationDevi.kt]: wat?

File: src/main/java/com/hedvig/botService/chat/OnboardingConversationDevi.kt [hunk 3]
```diff
@@ -1384,6 +1493,34 @@ constructor(
...
+                userContext.putUserData("{LOGIN_EMAIL}", trim)
+                m.body.text = trim
+                addToChat(m, userContext)
+                nxtMsg = if (trim == "apple@hedvig.com") {
```
Comment (palmfjord) [inline src/main/java/com/hedvig/botService/chat/OnboardingConversationDevi.kt]: i really don't think we should hard-code this

File: src/main/java/com/hedvig/botService/chat/OnboardingConversationDevi.kt [hunk 4]
```diff
@@ -1384,6 +1493,34 @@ constructor(
...
+                    MESSAGE_LOGIN_WITH_MAIL_ASK_PASSWORD
+                } else {
+                    MESSAGE_LOGIN_FAILED_WITH_MAIL
+                }
+            }
+            MESSAGE_LOGIN_WITH_MAIL_ASK_PASSWORD -> {
+                val trim = m.body.text.trim { it <= ' ' }
```
Comment (palmfjord) [inline src/main/java/com/hedvig/botService/chat/OnboardingConversationDevi.kt]: trim? isn't this rather something like `trimmedPassword`?

File: src/main/java/com/hedvig/botService/services/SessionManager.java [hunk 5]
```diff
@@ -119,9 +122,31 @@ public BankIdCollectResponse collect(String hid, String referenceToken) {
...
+      hid,
+      referenceToken,
+      (BankIdChat) conversationFactory.createConversation(OnboardingConversationDevi.class));
+  }
+
+  public boolean emailSign(String memberId){
+    UserContext uc =
+      userContextRepository
+        .findByMemberId(memberId)
+        .orElseThrow(() -> new ResourceNotFoundException("Could not find usercontext."));
```
Comment (palmfjord) [inline src/main/java/com/hedvig/botService/services/SessionManager.java]: maybe something about `"...for member" + memberId`

File: src/main/java/com/hedvig/botService/web/InternalUserDataController.java [hunk 6]
```diff
@@ -1,32 +1,38 @@
...
 
 @RestController
 @RequestMapping("/_/member/")
 public class InternalUserDataController {
 
   private static Logger log = LoggerFactory.getLogger(InternalUserDataController.class);
   private final SessionManager sessionManager;
+  private final UserContextRepository repository;
+
+  @Value("${hedvig.apple.user.memberId:00000}")
```
Comment (palmfjord) [inline src/main/java/com/hedvig/botService/web/InternalUserDataController.java]: should we even have a default value here?

File: src/main/java/com/hedvig/botService/chat/OnboardingConversationDevi.kt (hunk 3 above)
Comment (meletios) [reply src/main/java/com/hedvig/botService/chat/OnboardingConversationDevi.kt]: lowercase

File: src/main/java/com/hedvig/botService/chat/OnboardingConversationDevi.kt [hunk 7]
```diff
@@ -363,14 +369,133 @@ constructor(
...
+                        "message.bankid.autostart.respond", null,
+                        "bankid:///?autostarttoken={AUTOSTART_TOKEN}&redirect={LINK_URI}", null,
+                        false
+                    ),
+                    SelectOption("Jag är inte medlem", MESSAGE_ONBOARDINGSTART_ASK_EMAIL)
+                )
+            ) { body, uc, message ->
+                body.text = body.selectedItem.text
+                addToChat(message, uc)
+                val obd = uc.onBoardingData
```
Comment (palmfjord) [inline src/main/java/com/hedvig/botService/chat/OnboardingConversationDevi.kt]: but why

File: src/main/java/com/hedvig/botService/chat/OnboardingConversationDevi.kt [hunk 8]
```diff
@@ -1384,6 +1508,33 @@ constructor(
                 endConversation(userContext)
                 return
             }
+            MESSAGE_LOGIN_ASK_EMAIL -> {
+                val trimEmail = m.body.text.trim()
```
Comment (palmfjord) [inline src/main/java/com/hedvig/botService/chat/OnboardingConversationDevi.kt]: should we maybe `toLowerCase` it here already?

File: src/main/java/com/hedvig/botService/chat/OnboardingConversationDevi.kt [hunk 9]
```diff
@@ -1384,6 +1508,33 @@ constructor(
...
+                    MESSAGE_LOGIN_WITH_MAIL_ASK_PASSWORD
+                } else {
+                    MESSAGE_LOGIN_FAILED_WITH_MAIL
+                }
+            }
+            MESSAGE_LOGIN_WITH_MAIL_ASK_PASSWORD -> {
+                val trimmedPwd = m.body.text.trim()
+                m.body.text = "*****"
+                addToChat(m, userContext)
+                if (trimmedPwd.toLowerCase() == APPLE_USER_PASSWORD.toLowerCase()) {
```
Comment (palmfjord) [inline src/main/java/com/hedvig/botService/chat/OnboardingConversationDevi.kt]: Passwords _should_ be case sensitive and not trimmed imo

File: src/main/java/com/hedvig/botService/services/SessionManager.java [hunk 10]
```diff
@@ -52,6 +51,10 @@
   private static final String LINK_URI_KEY = "{{LINK_URI}";
   private static final String LINK_URI_VALUE = "hedvig://+";
 
+  private static final String LOGIN = "{LOGIN}";
+  private static final String LOGIN_MAIL = "{LOGIN_EMAIL}";
+  private static final String LOGIN_MAIL_MEMBER_ID = "12345"; //TODO: Change this
```
Comment (palmfjord) [inline src/main/java/com/hedvig/botService/services/SessionManager.java]: 🙈

File: src/main/java/com/hedvig/botService/services/SessionManager.java [hunk 11]
```diff
@@ -119,9 +122,31 @@ public BankIdCollectResponse collect(String hid, String referenceToken) {
...
+    UserContext uc =
+      userContextRepository
+        .findByMemberId(memberId)
+        .orElseThrow(() -> new ResourceNotFoundException("Could not find usercontext for member: "+ memberId));
+
+    if (uc.getDataEntry(LOGIN).equalsIgnoreCase("true") && uc.getDataEntry(LOGIN_MAIL).equalsIgnoreCase("apple@hedvig.com")) {
+
+      UserContext newUc =
+        userContextRepository
+          .findByMemberId("APPLE")
```
Comment (palmfjord) [inline src/main/java/com/hedvig/botService/services/SessionManager.java]: should this really be `APPLE`? Sholudn't it be whatever we configure the apple member id to be?

File: src/main/java/com/hedvig/botService/services/SessionManager.java [hunk 12]
```diff
@@ -119,9 +122,31 @@ public BankIdCollectResponse collect(String hid, String referenceToken) {
...
+          .orElseThrow(() -> new ResourceNotFoundException("Could not fina userContext."));
```
Comment (palmfjord) [inline src/main/java/com/hedvig/botService/services/SessionManager.java]: typo

File: src/main/java/com/hedvig/botService/services/SessionManager.java [hunk 13]
```diff
@@ -119,9 +122,31 @@ public BankIdCollectResponse collect(String hid, String referenceToken) {
...
+      (BankIdChat) conversationFactory.createConversation(OnboardingConversationDevi.class));
+  }
+
+  public boolean emailSign(String memberId){
+    UserContext uc =
+      userContextRepository
+        .findByMemberId(memberId)
+        .orElseThrow(() -> new ResourceNotFoundException("Could not find usercontext for member: "+ memberId));
+
+    if (uc.getDataEntry(LOGIN).equalsIgnoreCase("true") && uc.getDataEntry(LOGIN_MAIL).equalsIgnoreCase("apple@hedvig.com")) {
```
Comment (palmfjord) [inline src/main/java/com/hedvig/botService/services/SessionManager.java]: configurable no?

File: src/main/java/com/hedvig/botService/web/AppleController.java [hunk 14]
```diff
@@ -0,0 +1,89 @@
...
+import org.springframework.http.ResponseEntity;
+import org.springframework.web.bind.annotation.PostMapping;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RestController;
+
+import java.util.Optional;
+
+@RestController
+@Slf4j
+@RequestMapping("i")
```
Comment (palmfjord) [inline src/main/java/com/hedvig/botService/web/AppleController.java]: `i`?

File: src/main/java/com/hedvig/botService/web/AppleController.java [hunk 15]
```diff
@@ -0,0 +1,89 @@
...
+    MemberService memberService,
+    ProductPricingService productPricingService, UserContextService userContextService) {
+    this.memberService = memberService;
+    this.productPricingService = productPricingService;
+    this.userContextService = userContextService;
+  }
+
+  @PostMapping("initAppleUser")
+  ResponseEntity<?> initializeAppleUser() {
+    log.info("Initializing Apple User, in memory of Steve Jobs!");
```
Comment (palmfjord) [inline src/main/java/com/hedvig/botService/web/AppleController.java]: ❤️

File: src/main/java/com/hedvig/botService/web/HedvigController.java [hunk 16]
```diff
@@ -78,4 +75,16 @@ public HedvigController(SessionManager sessions) {
...
   }
+
+  @PostMapping("emailSign")
+  ResponseEntity<?> emailSign(@RequestHeader(value = "hedvig.token") String memberId) {
+    log.info("Email login for member: {}", memberId);
+
+    boolean isEmailSignSuccessful = this.sessionManager.emailSign(memberId);
+
+    if (isEmailSignSuccessful) {
+      return ResponseEntity.ok().header("Hedvig.Id", "").build(); // TODO: APPLEUSERID
```
Comment (palmfjord) [inline src/main/java/com/hedvig/botService/web/HedvigController.java]: 🔍

File: src/main/java/com/hedvig/botService/chat/OnboardingConversationDevi.kt [hunk 17]
```diff
@@ -1632,6 +1783,16 @@ constructor(
         }
     }
 
+    fun emailLoginComplete(uc: UserContext) {
+        when {
+            uc.onBoardingData.userHasSigned!! -> {
```
Comment (palmfjord) [inline src/main/java/com/hedvig/botService/chat/OnboardingConversationDevi.kt]: couldn't this produce an npe quite easily? Wouldn't it be better do do something like `uc.onBoardingData.userHasSigned ?: false -> {`?

File: src/main/java/com/hedvig/botService/services/SessionManager.java (hunk 13 above)
Comment (palmfjord) [reply src/main/java/com/hedvig/botService/services/SessionManager.java]: also maybe we should check the password?

File: src/main/java/com/hedvig/botService/chat/OnboardingConversationDevi.kt [hunk 18]
```diff
@@ -363,14 +369,133 @@ constructor(
...
+                    }
+                    MESSAGE_ONBOARDINGSTART_ASK_EMAIL -> {
+                        uc.putUserData(LOGIN, "false")
+                    }
+                }
+                body.selectedItem.value
+            })
+        setupBankidErrorHandlers(MESSAGE_LOGIN_FAILED_WITH_MAIL)
+
+        this.createChatMessage(MESSAGE_LOGIN_WITH_MAIL,
```
Comment (palmfjord) [inline src/main/java/com/hedvig/botService/chat/OnboardingConversationDevi.kt]: `email` no? (all `mail`s)

File: src/main/java/com/hedvig/botService/chat/OnboardingConversationDevi.kt (hunk 7 above)
Comment (meletios) [reply src/main/java/com/hedvig/botService/chat/OnboardingConversationDevi.kt]: but why but why

File: src/main/java/com/hedvig/botService/web/AppleController.java (hunk 14 above)
Comment (meletios) [reply src/main/java/com/hedvig/botService/web/AppleController.java]: like internal?

**Pull request merged**